third_party/move/move-vm/runtime/src/storage/environment.rs

```diff
@@ -0,0 +1,315 @@
+    /// there are duplicated natives, panics.
+    pub fn new(
+        natives: impl IntoIterator<Item = (AccountAddress, Identifier, Identifier, NativeFunction)>,
+    ) -> Self {
+        let vm_config = VMConfig {
+            // Keep the paranoid mode on as we most likely want this for tests.
+            paranoid_type_checks: true,
+            ..VMConfig::default()
+        };
+        Self::new_with_config(natives, vm_config)
+    }
+
+    /// Creates a new runtime environment with native functions and VM configurations. If there are
+    /// duplicated natives, panics.
+    pub fn new_with_config(
+        natives: impl IntoIterator<Item = (AccountAddress, Identifier, Identifier, NativeFunction)>,
+        vm_config: VMConfig,
+    ) -> Self {
+        let natives = NativeFunctions::new(natives)
+            .unwrap_or_else(|e| panic!("Failed to create native functions: {}", e));
+        Self {
+            vm_config,
+            natives,
+            verifier_extension: None,
+            struct_name_index_map: StructNameIndexMap::empty(),
+            ty_cache: RwLock::new(TypeCache::empty()),
+        }
+    }
+
+    /// Returns the config currently used by this runtime environment.
+    pub fn vm_config(&self) -> &VMConfig {
+        &self.vm_config
+    }
+
+    /// Enables delayed field optimization for this environment.
+    pub fn enable_delayed_field_optimization(&mut self) {
+        self.vm_config.delayed_field_optimization_enabled = true;
+    }
+
+    /// Creates a partially verified compiled script by running:
+    ///   1. Move bytecode verifier,
+    ///   2. Verifier extension, if provided.
+    pub fn build_partially_verified_script(
+        &self,
+        compiled_script: Arc<CompiledScript>,
+    ) -> VMResult<PartiallyVerifiedScript> {
+        move_bytecode_verifier::verify_script_with_config(
+            &self.vm_config().verifier_config,
+            compiled_script.as_ref(),
+        )?;
+        if let Some(verifier) = &self.verifier_extension {
+            verifier.verify_script(compiled_script.as_ref())?;
+        }
+        Ok(PartiallyVerifiedScript(compiled_script))
+    }
+
+    /// Creates a fully verified script by running dependency verification pass. The caller must
+    /// provide verified module dependencies.
+    pub fn build_verified_script(
+        &self,
+        partially_verified_script: PartiallyVerifiedScript,
+        immediate_dependencies: &[Arc<Module>],
+    ) -> VMResult<Script> {
+        dependencies::verify_script(
+            partially_verified_script.0.as_ref(),
+            immediate_dependencies.iter().map(|m| m.module()),
+        )?;
+        Script::new(partially_verified_script.0, self.struct_name_index_map())
+            .map_err(|e| e.finish(Location::Script))
+    }
+
+    /// Creates a partially verified compiled module by running:
+    ///   1. Move bytecode verifier,
+    ///   2. Verifier extension, if provided.
+    pub fn build_partially_verified_module(
```

Comment:
Is the problem that we need to recursively run dependency verification first and then build the module itself, yet we want to control the traversal logic at the aptos side rather than the move side? In that case should we rename the paritially verified to LocalVerifiedScript/Module?

Reply:
Exactly! Will rename.